### PR TITLE
fix Issue 22275 - importC: Error: incompatible types for '(dest) !is (buf)': 'char*' and 'char[1]'

### PR DESCRIPTION
**Pull request by ibuclaw**
Move logic to `typeCombine`, add  handling for array types to saturate to pointers.

### CODE REVIEW

File: src/dmd/dcast.d [hunk 1]
```diff
@@ -3453,6 +3453,40 @@ Expression typeCombine(BinExp be, Scope* sc)
...
+            if (auto ie = be.e1.isIntegerExp())
+            {
+                if (ie.toInteger() == 0)
+                    be.e1 = new NullExp(ie.loc, t2);
+            }
+            else if (auto ie = be.e2.isIntegerExp())
+            {
+                if (ie.toInteger() == 0)
+                    be.e2 = new NullExp(ie.loc, t1);
+            }
```
Comment (WalterBright) [inline src/dmd/dcast.d:3486]: This code is only partially correct. For example, this should compile:
```
int *p;
p = 1;
```
To correct this code, use the function `cimplictConvto()`.

File: src/dmd/dcast.d [hunk 2]
```diff
@@ -3453,6 +3453,40 @@ Expression typeCombine(BinExp be, Scope* sc)
...
+        if (t1.ty == Tsarray || t1.ty == Tarray)
+        {
+            be.e1 = new DotIdExp(be.loc, be.e1, Id.ptr).expressionSemantic(sc);
+            t1 = be.e1.type.toBasetype();
+        }
+        if (t2.ty == Tsarray || t2.ty == Tarray)
+        {
+            be.e2 = new DotIdExp(be.loc, be.e2, Id.ptr).expressionSemantic(sc);
+            t2 = be.e2.type.toBasetype();
+        }
```
Comment (WalterBright) [inline src/dmd/dcast.d:3469]: Use `arrayFuncConv()` instead, which already does this (and also handles function conversion to pointer).

File: test/compilable/testcstuff2.c [hunk 3]
```diff
@@ -404,10 +404,22 @@ int test22245()
...
 {
-  if (buf == 0)
-    return;
-  if (0 == buf)
-    return;
+    if (buf == 0)
+        return;
+    if (0 == buf)
+        return;
+}
```
Comment (WalterBright) [inline test/compilable/testcstuff2.c:411]: Please add test case:
```
char *p;
p = x ? p : 1;
p = x ? 1 : p;
```

**Review by WalterBright — CHANGES_REQUESTED** [review]
see comments

**Comment by WalterBright** [comment]
@ibuclaw ping ?

**Comment by RazvanN7** [comment]
@ibuclaw Please apply the review feedback and let's get this in.

**Pull request closed without merging**